ecl/eclcc/WORKUNITS.rst

```diff
@@ -0,0 +1,862 @@
+                set3->addRange(v4,v4);
+            }
+            irc->append(createKeySegmentMonitor(false, set3.getClear(), 0, 40));
+            irc->append(createWildKeySegmentMonitor(40, 80));
+        }
+        virtual size32_t transform(ARowBuilder & crSelf, const void * _left) {
+            crSelf.getSelf();
+            unsigned char * left = (unsigned char *)_left;
+            memcpy(crSelf.row() + 0U,left + 0U,120U);
+            return 120U;
+        }
+    };
+
+Some of the methods to highlight are:
+
+a) onCreate() - common to all activities.  It is called by the engine when the helper is first created,
+   and allows the helper to cache information which doesn't change - in this case the name that is
+   being searched for.
+b) getFileName() - determines the name of the index being read.
+c) createSegmentMonitors() - defines which columns are being filtered, and what values to match against.
+d) transform() - create the record to return from the activity.  It controls which columns should be included
+   from the index row in the output. (In this case all.)
+
+Executing the graph
+-------------------
+
+To execute a graph, the engine walks the activities in the graph xml and creates in memory a graph of implementation
+activities:
+
+For each activity, the name of the helper factory is calculated from the activity number (e.g., fAc2 for
+activity 2), that function is imported from the loaded dll, and then called to create an instance of the generated helper
+class - in this case cAc2.
+
+The engine creates an instance of the class for implementing that kind of activity.  (E.g., activity 2 
+has a _kind of 77 - which corresponds to TAKindexread, so roxie will create an instance of CRoxieServerIndexReadActivity)
+and pass the previously created helper object (which must implement IHThorIndexReadArg) to the constructor.  Some
+activity implementations may also extract other information from the xml for the activity - e.g., hints.  Once all the
+activities are created the edge information is used to link inputs to outputs and add other dependencies.  
+
+Note: Any subgraph that is marked with <att name="rootGraph" value="1"/> is a root subgraph.  An activity within
+a subgraph that has no outputs is called a sink (activities without inputs are called sources).
+
+Executing a graph involves executing all the root subgraphs that it contains.  All
+dependencies of the activities within the subgraph must be executed before a subgraph is executed.  
+To execute a subgraph, the engine executes each of the sink activities
+on separate threads, and then waits for each of the threads to complete.  
+Each sink activity lazily pulls input rows on demand from activities further up the graph, processes than and returns
+when complete.
+
+(If you examine the code you will find the implementation for processing dependencies is more fine grained, to ensure
+IF datasets, OUPUT(,UPDATE) and other ECL constructs are executed correctly.)
```

Comment:
I know this is only being mentioned as a *note*, of sorts, but it should perhaps be explained.